### DISCUSSION TITLE
Adding upper and lower limits to a time series and styling of limits.

### DISCUSSION THREAD

**Discussion by Oliver-Loeffler** [GSI-CS-CO/chart-fx#404] [Q&A]
Hello!

Chart-fx is really an awesome project and it works great - I am kind of new to it and still learning the way how it is assembled.
There 2 questions came to my mind, hopefully you can help.

* What is the best (or intended) way to add upper and lower limits to e.g. time series charts? The limits may vary over time.
* Is there a separate API for that or is that just another dataset?
* If it is another data set, what is the desired way of styling data sets? 

Creating a custom dataset implementation and also a custom renderer might be the way to go.

Thanks!

**Comment by ennerf** [comment]
Can you be more specific about the use case? Do you want to limit the time axis or the value axis?

I'm not a maintainer, but I've come across some use cases where I e.g. needed to limit a time series to the last point minus N seconds. That was pretty easy to solve with a custom dataset implementation that updates the axis limits as needed. Something like

```java
private void recomputeLimitsInternal() {
        AxisDescription xAxis = getAxisDescription(0);
        AxisDescription yAxis = getAxisDescription(1);

       xAxis.set(minTime, maxTime);
       yAxis.set(minValue, maxValue); // or call yAxis.clear(), and yAxis.add(value) for all included values
}
```

A custom renderer shouldn't be necessary unless you need to optimize performance.

**Reply by Oliver-Loeffler** [reply]
The use case is, to inidicate if a value exceeds or is below a certain limit. So the limits shal apply for the value axis. The actual limits applied to the dataset may change over time. So its not just a straight line crossing the value axis.
Also the limits shall be drawn alongside the data.

**Reply by RalphSteinhagen** [reply]
Hi @Oliver-Loeffler thanks for your interest. 

The solution @ennerf provided is good and valid if you want to modify the default auto-ranging and have a custom axis range. For example, one that does not necessarily depend one-to-one on the data ranges of your datasets and/or if you have already have computed the minimum/maximum data range by other means. Alternatively, you can also use static (i.e. non auto-ranging) axis ranges or modify the axis ranges dynamically through their properties.

> [..] to inidicate if a value exceeds or is below a certain limit [..]

For confidence limits (aka. 'error bars', 'error surfaces', 'whiskers', etc.) and exclusion-type limits the following snippet may be useful:

```Java
public class LimitsSample extends Application {
    private static final int N_SAMPLES = 100;
    private static final int WIDTH = 1200;
    private static final int HEIGHT = 600;

    @Override
    public void start(final Stage primaryStage) {
        final DefaultNumericAxis xAxis = new DefaultNumericAxis("time", "s");
        final DefaultNumericAxis yAxis = new DefaultNumericAxis("y-value", "a.u.");
        final XYChart chart = new XYChart(xAxis, yAxis);
        chart.getPlugins().add(new Zoomer());
        chart.getPlugins().add(new EditAxis());
        ErrorDataSetRenderer rendererValue = (ErrorDataSetRenderer) chart.getRenderers().get(0);
        rendererValue.setDrawMarker(true);
        ErrorDataSetRenderer rendererLimits = new ErrorDataSetRenderer();
        rendererLimits.setErrorType(ErrorStyle.ERRORSURFACE);
        rendererLimits.setDrawMarker(false);
        chart.getRenderers().add(rendererLimits);

        DoubleErrorDataSet upperLimit = new DoubleErrorDataSet("upper limit");
        upperLimit.setStyle("showInLegend=false; strokeColor=darkgreen");
        DoubleErrorDataSet lowerLimit = new DoubleErrorDataSet("lower limit");
        lowerLimit.setStyle("showInLegend=false; strokeColor=darkred");
        DoubleErrorDataSet dataSet = new DoubleErrorDataSet("some dataset");
        dataSet.setStyle("markerSize=3; markerColor=violet; markerType=diamond");
        rendererValue.getDatasets().add(dataSet);
        rendererLimits.getDatasets().addAll(upperLimit, lowerLimit);

        for (int i = 0; i < N_SAMPLES; i++) {
            final double time = i * 0.01;
            final double value = MathBaseFast.sin(2.0 * Math.PI * time);
            final double confidenceInterval = 0.2 * MathBaseFast.cos(2.0 * Math.PI * time);

            dataSet.add(time, value, confidenceInterval, confidenceInterval);
            upperLimit.add(time, value + 2, 0, Double.POSITIVE_INFINITY);
            lowerLimit.add(time, value - 2, Double.POSITIVE_INFINITY, 0);
        }

        final Scene scene = new Scene(chart, WIDTH, HEIGHT);

        primaryStage.setTitle(LimitsSample.class.getSimpleName());
        primaryStage.setOnCloseRequest(evt -> Platform.exit());
        primaryStage.setScene(scene);
        primaryStage.show();
    }
}
```
yielding the following:
![image](https://user-images.githubusercontent.com/46007894/115120928-5a1f6700-9fb0-11eb-8bf3-e38d205fb4a4.png)

Does this help?

> If it is another data set, what is the desired way of styling data sets?

There are two options: you can style 
a) specific datasets (as shown above) or 
b) modify the standard styles that are applied to all datasets.

Feel free to also post graphs or pictures of what you would like to achieve... May I ask the domain you are working in?

**Reply by Oliver-Loeffler** [reply]
Thank you for your help @ennerf & @RalphSteinhagen - I will try that out. You helped me a lot. To go that way was not abvious for me but this will do the job. Also the styling question as answered, so that helps as know I know where to dive in. I will prepare an example. Well, it's going to be interesting as there will be multiple sets of limits, kind of control limits, alarm limits, reasonable limits etc.

Where are the standard styles for all datasets defined?

I'm working in metrology process engineering, so tracking metrology machine parameters to detect faults and also to determine measurement uncertainties. Currently I'm exploring chart-fx to be used for showing data in statistical process control context.

**Reply by RalphSteinhagen** [reply]
> Where are the standard styles for all datasets defined?

The default colour schemes are defined in [DefaultRenderColorScheme](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/utils/DefaultRenderColorScheme.java).
